src/snowflake/snowpark/functions.py

```diff
@@ -2289,6 +2289,38 @@ def substring(
     return builtin("substring")(s, p, length)
 
 
+def substring_index(str: ColumnOrName, delim: str, count: int) -> Column:
```

Comment:
strtok_to_array doesn't see working right to me, it errors when I call df. strtok_to_array("s", "."). Actually i don't think we should make it ColumnOrName, the delimiter is really just a string.
This API is currently compatible with pyspark (the same signature), it will probably fail if customer directly moves from pyspark.

Reply:
yea, I think `delim` in `strtok_to_array` should be a string, like what we have in `listagg`...

Reply:
I suggest not using `str` as the name of argument as it is reserved keyword for python. Just using `c` is fine

Reply:
So we have a bug with `strtok_to_array `?

Reply:
I just tried, this in Snowflake works `select STRTOK_TO_ARRAY(a, b) from (select 'a.b' as a, '.' as b);`. That means the delim can be a column in Snowflake SQL. In our APIs, usually we accept a string as a column name when that parameter can be a column.
`listagg` document explicitly says its delimiter must be a constant. So it's different from strtok_to_array.

Reply:
so strtok_to_array in snowflake allows three cases 1) a column, 2) a single delimiter, like '.' 3) a set of delimiter, like '.@,'. However, today's implementation in snowpark seems only handle columns, when i call it with df.select(strtok_to_array("s", ".")).collect() , it will throw an error  invalid identifier '"."', because we handle it always like a column. This seems like a restriction to me because this requires customer always project the delimiter as a column, but I believe most of the use case is just using a string. To handle it correctly, we will need to do a check first to see if the delimiter corresponds to a column, if not, simply treat it as a regular string.

substring_index currently do not have an direct support in snowflake, i don't think i against to extend the support to handle column, however, it should handle the case of string correctly to make sure we can support case that pyspark supports first.

Reply:
```
I suggest not using str as the name of argument as it is reserved keyword for python. Just using c is fine
```
Sure. I updated it to text which seems another commonly used function parameter  name in functions.py